packages/@react-stately/steplist/src/useStepListState.ts

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2023 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import {AriaStepListProps, StepListState} from '@react-types/steplist';
+import {Collection, Node} from '@react-types/shared';
+import {Key, useEffect, useMemo} from 'react';
+import {useControlledState} from '@react-stately/utils';
+import {useSingleSelectListState} from '@react-stately/list';
+
+export function useStepListState<T extends object>(props: AriaStepListProps<T>): StepListState<T> {
+  let state = useSingleSelectListState<T>(props);
+
+  let [lastCompletedStep, setLastCompletedStep] = useControlledState<Key>(props.lastCompletedStep, props.defaultLastCompletedStep, props.onLastCompletedStepChange);
+  const {setSelectedKey: realSetSelectedKey, selectedKey, collection} = state;
+  const {prevKeyMap, indexMap} = useMemo(() => buildKeysMaps(collection), [collection]);
```

Comment:
is this just for keeping track of the currently selected key and then marking all before it as completed?
instead of building this map, could we just use collection.getKeyBefore?
could just keep a Set of all keys before the selected one, might be a little clearer?
instead of `indexMap.get(step) <= indexMap.get(lastCompletedStep);` you could just write `completed.has(key)`

Reply:
also, i'm not quite following why it's called prevKeyMap, it's current?

ooo, I just figured out this means, previous key... map, not previous... key map. if we're going to keep it, then lets rename it. mapOfPreviousKeys? keysLinkedList?